cli.go

```diff
@@ -0,0 +1,212 @@
+		var response string
+
+		query := req.URL.Query()
+		code := query.Get("code")
+		state := query.Get("state")
+		data := map[string][]string{
+			"code":  {code},
+			"state": {state},
+		}
+
+		secret, err := c.Logical().ReadWithData(fmt.Sprintf("auth/%s/oidc/callback", mount), data)
+		if err != nil {
+			summary, detail := parseError(err)
+			response = errorHTML(summary, detail)
+		} else {
+			response = successHTML
+		}
+
+		w.Write([]byte(response))
+		doneCh <- loginResp{secret, err}
+	})
+
+	listener, err := net.Listen("tcp", ":"+port)
+	if err != nil {
+		return nil, err
+	}
+
+	// Open the default browser to the callback URL.
+	fmt.Fprintf(os.Stderr, "Complete the login via your OIDC provider. Launching browser to:\n\n    %s\n\n\n", authURL)
+	if err := openURL(authURL); err != nil {
+		fmt.Fprintf(os.Stderr, "Error attempting to automatically open browser: '%s'.\nPlease visit the authorization URL manually.", err)
+	}
+
+	// Start local server
+	go func() {
+		err := http.Serve(listener, nil)
+		if err != nil && err != http.ErrServerClosed {
+			doneCh <- loginResp{nil, err}
+		}
+	}()
+
+	// Wait for either the callback to finish or SIGINT to be received
+	select {
+	case s := <-doneCh:
+		return s.secret, s.err
+	case <-sigintCh:
+		return nil, errors.New("interrupted")
+	}
+}
+
+func fetchAuthURL(c *api.Client, role, mount, port string) (string, error) {
+	data := map[string]interface{}{
+		"role":         role,
+		"redirect_uri": fmt.Sprintf("http://localhost:%s/ui/vault/auth/%s/oidc/callback", port, mount),
+	}
+
+	secret, err := c.Logical().Write(fmt.Sprintf("auth/%s/oidc/auth_url", mount), data)
+	if err != nil {
+		return "", err
+	}
+
+	authURL := secret.Data["auth_url"].(string)
+	if authURL == "" {
+		return "", errors.New(fmt.Sprintf("Unable to authorize role %q. Check Vault logs for more information.", role))
+	}
+
+	return authURL, nil
+}
+
+// openURL opens the specified URL in the default browser of the user.
+// Source: https://stackoverflow.com/a/39324149/453290
+func openURL(url string) error {
+	var cmd string
+	var args []string
+
+	switch runtime.GOOS {
+	case "windows":
+		cmd = "cmd"
+		args = []string{"/c", "start"}
+	case "darwin":
+		cmd = "open"
+	default: // "linux", "freebsd", "openbsd", "netbsd"
+		cmd = "xdg-open"
+	}
+	args = append(args, url)
+	return exec.Command(cmd, args...).Start()
+}
+
+// parseError converts error from the API into summary and detailed portions.
+// This is used to present a nicer UI by splitting up *known* prefix sentences
+// from the rest of the text. e.g.
+//
+//    "No response from provider. Gateway timeout from upstream proxy."
+//
+// becomes:
+//
+//    "No response from provider.", "Gateway timeout from upstream proxy."
+func parseError(err error) (string, string) {
+	headers := []string{errNoResponse, errLoginFailed, errTokenVerification}
+	summary := "Login error"
+	detail := ""
+
+	re := regexp.MustCompile(`(?s)Errors:.*\* *(.*)`)
```

Comment:
Usually must compiles are put into init() functions so that you don't get somewhere along the path of a request and *then* panic. I'd suggest using that pattern here.

Reply:
Moved regex compilation outside of the request path.